File: pdv-python/pdv/handlers/introspection.py
"""
pdv.handlers.introspection — introspection and path-resolution handlers
used by the MCP server.

Handles:
- ``pdv.help``: resolve a dotted symbol to a live Python object and return
  its kind, signature, docstring, and (optionally) source.
- ``pdv.tree.resolve_path``: translate between a dot-delimited tree path and
  an absolute filesystem path, in either direction.

These handlers exist to give the PDV MCP server read-only insight into the
kernel namespace and the file-backed tree nodes without needing to execute
arbitrary code.

See Also
--------
ARCHITECTURE.md §15 (MCP server)
"""

from __future__ import annotations

import importlib
import inspect
import os
from typing import Any

from pdv.handlers import register


def _resolve_symbol(symbol: str) -> Any:
    """Resolve a dotted symbol string to a live Python object.

    Splits *symbol* on ``.``. The head is looked up first in the kernel
    user namespace, and failing that imported as a module. The remaining
    parts are walked with :func:`getattr`.

    Parameters
    ----------
    symbol : str
        A dotted symbol, e.g. ``'pdv'``, ``'pdv.add_file'``,
        ``'numpy.ndarray'``, or a bare user-namespace variable name.

    Returns
    -------
    Any
        The resolved live object.

    Raises
    ------
    KeyError, AttributeError, ImportError, ValueError
        If the symbol cannot be resolved at any step.
    """
    from pdv.comms import get_ip  # noqa: PLC0415

    if not symbol:
        raise ValueError("symbol must be a non-empty string")

    parts = symbol.split(".")
    head, *rest = parts

    ip = get_ip()
    user_ns = ip.user_ns if ip is not None else {}

    if head in user_ns:
        obj: Any = user_ns[head]
    else:
        # Not in the namespace — try importing it as a module. Walk the
        # longest importable dotted prefix so e.g. ``numpy.linalg.norm``
        # imports ``numpy.linalg`` and then getattrs ``norm``.
        obj = importlib.import_module(head)
        while rest:
            try:
                obj = importlib.import_module(f"{head}.{rest[0]}")
            except ImportError:
                break
            head = f"{head}.{rest[0]}"
            rest = rest[1:]

    for attr in rest:
        obj = getattr(obj, attr)
    return obj


def _object_kind(obj: Any) -> str:
    """Return a coarse string label describing what kind of object *obj* is.

    Parameters
    ----------
    obj : Any
        The object to classify.

    Returns
    -------
    str
        One of ``'module'``, ``'class'``, ``'method'``, ``'function'``,
        or ``'object'``.
    """
    if inspect.ismodule(obj):
        return "module"
    if inspect.isclass(obj):
        return "class"
    if inspect.ismethod(obj):
        return "method"
    if inspect.isfunction(obj) or inspect.isbuiltin(obj) or inspect.isroutine(obj):
        return "function"
    return "object"


def handle_help(msg: dict) -> None:
    """Handle the ``pdv.help`` message.

    Resolves a dotted symbol to a live Python object and returns
    introspection metadata for it.

    Expected payload
    ----------------
    .. code-block:: json

        {
            "symbol": "pdv.add_file",
            "include_source": false
        }

    ``include_source`` is optional and defaults to ``false``.

    Response payload (``pdv.help.response``)
    ----------------------------------------
    .. code-block:: json

        {
            "symbol": "pdv.add_file",
            "kind": "function",
            "signature": "(source_path: str) -> PDVFile",
            "doc": "Import an arbitrary file ...",
            "source": null
        }

    ``signature``, ``doc``, and ``source`` are ``null`` when not
    introspectable. ``source`` is only ever populated when
    ``include_source`` is true.

    Parameters
    ----------
    msg : dict
        Parsed PDV message envelope.
    """
    from pdv.comms import send_error, send_message  # noqa: PLC0415

    msg_id = msg.get("msg_id")
    payload = msg.get("payload", {})
    symbol = payload.get("symbol", "")
    include_source = bool(payload.get("include_source", False))

    if not symbol:
        send_error(
            "pdv.help.response",
            "introspection.missing_symbol",
            "symbol is required in pdv.help payload",
            in_reply_to=msg_id,
        )
        return

    try:
        obj = _resolve_symbol(symbol)
    except (KeyError, AttributeError, ImportError, ValueError, TypeError) as exc:
        send_error(
            "pdv.help.response",
            "introspection.symbol_not_found",
            f"Could not resolve symbol '{symbol}': {exc}",
            in_reply_to=msg_id,
        )
        return

    kind = _object_kind(obj)

    signature: str | None = None
    try:
        signature = str(inspect.signature(obj))
    except (TypeError, ValueError):
        signature = None

    doc: str | None = None
    try:
        doc = inspect.getdoc(obj)
    except Exception:  # noqa: BLE001
        doc = None

    source: str | None = None
    if include_source:
        try:
            source = inspect.getsource(obj)
        except (OSError, TypeError):
            source = None

    send_message(
        "pdv.help.response",
        {
            "symbol": symbol,
            "kind": kind,
            "signature": signature,
            "doc": doc,
            "source": source,
        },
        in_reply_to=msg_id,
    )


def _collect_file_nodes(
    value: Any,
    tree_path: str,
    out: list[tuple[str, Any]],
) -> None:
    """Recursively collect ``(tree_path, PDVFile)`` pairs from a container.

    Walks dicts (including PDVTree) accumulating every :class:`PDVFile`
    node it encounters, keyed by its dot-delimited tree path.

    Parameters
    ----------
    value : Any
        The current node value being walked.
    tree_path : str
        Dot-delimited path of *value* within the root tree.
    out : list[tuple[str, PDVFile]]
        Accumulator that collected pairs are appended to.
    """
    from pdv.tree import PDVFile  # noqa: PLC0415

    if isinstance(value, PDVFile):
        out.append((tree_path, value))
        return
    if isinstance(value, dict):
        for key in list(dict.keys(value)):
            child = dict.__getitem__(value, key)
            child_path = f"{tree_path}.{key}" if tree_path else key
            _collect_file_nodes(child, child_path, out)


def _extract_uuid_segment(abs_path: str) -> str | None:
    """Extract the ``<uuid>`` segment from a path under a ``tree/`` directory.

    A UUID-based file lives at ``<working_dir>/tree/<uuid>/<filename>``.
    If *abs_path* points at or inside such a ``tree/<uuid>/`` directory,
    return the ``<uuid>`` segment; otherwise return ``None``.

    Parameters
    ----------
    abs_path : str
        An absolute filesystem path.

    Returns
    -------
    str or None
        The UUID path segment, or ``None`` if *abs_path* is not inside a
        ``tree/<uuid>/`` directory.
    """
    parts = os.path.normpath(abs_path).split(os.sep)
    for index, part in enumerate(parts):
        if part == "tree" and index + 1 < len(parts):
            candidate = parts[index + 1]
            if candidate:
                return candidate
    return None


def handle_resolve_path(msg: dict) -> None:
    """Handle the ``pdv.tree.resolve_path`` message.

    Bidirectional path translation between a dot-delimited tree path and
    an absolute filesystem path.

    Expected payload
    ----------------
    .. code-block:: json

        { "path": "scripts.analysis.fit_model" }

    Direction is detected from *path*: an absolute filesystem path is
    treated as a reverse lookup (file → tree path(s)); anything else is a
    forward lookup (tree path → file path).

    Response payload (``pdv.tree.resolve_path.response``)
    -----------------------------------------------------
    .. code-block:: json

        {
            "input": "scripts.analysis.fit_model",
            "tree_paths": ["scripts.analysis.fit_model"],
            "file_path": "/abs/path/tree/<uuid>/fit_model.py"
        }

    ``file_path`` is ``null`` when the tree node is not file-backed (or,
    for reverse lookups, when no file path can be resolved). ``tree_paths``
    may contain multiple entries when several nodes share a UUID (copies).

    Parameters
    ----------
    msg : dict
        Parsed PDV message envelope.
    """
    from pdv.comms import get_pdv_tree, send_error, send_message  # noqa: PLC0415
    from pdv.tree import PDVFile  # noqa: PLC0415

    msg_id = msg.get("msg_id")
    payload = msg.get("payload", {})
    path = payload.get("path", "")

    if not path:
        send_error(
            "pdv.tree.resolve_path.response",
            "introspection.missing_path",
            "path is required in pdv.tree.resolve_path payload",
            in_reply_to=msg_id,
        )
        return

    tree = get_pdv_tree()
    if tree is None:
        send_error(
            "pdv.tree.resolve_path.response",
            "tree.no_tree",
            "PDVTree is not initialized",
            in_reply_to=msg_id,
        )
        return

    working_dir = getattr(tree, "_working_dir", None)

    if os.path.isabs(path):
        # Reverse: filesystem path -> tree path(s).
        target_real = os.path.realpath(path)
        target_uuid = _extract_uuid_segment(target_real)

        file_nodes: list[tuple[str, Any]] = []
        _collect_file_nodes(tree, "", file_nodes)

        tree_paths: list[str] = []
        for node_path, node in file_nodes:
            matched = False
            if target_uuid is not None and node.uuid == target_uuid:
                matched = True
            if not matched and working_dir:
                try:
                    node_real = os.path.realpath(node.resolve_path(working_dir))
                except (RuntimeError, ValueError):
                    node_real = None
                if node_real is not None and node_real == target_real:
                    matched = True
            if matched:
                tree_paths.append(node_path)

        send_message(
            "pdv.tree.resolve_path.response",
            {
                "input": path,
                "tree_paths": tree_paths,
                "file_path": target_real,
            },
            in_reply_to=msg_id,
        )
        return

    # Forward: tree path -> filesystem path.
    if path not in tree:
        send_error(
            "pdv.tree.resolve_path.response",
            "tree.path_not_found",
            f"No node at tree path: '{path}'",
            in_reply_to=msg_id,
        )
        return

    node = tree[path]
    file_path: str | None = None
    if isinstance(node, PDVFile):
        try:
            file_path = node.resolve_path(working_dir)
        except (RuntimeError, ValueError):
            file_path = None

    send_message(
        "pdv.tree.resolve_path.response",
        {
            "input": path,
            "tree_paths": [path],
            "file_path": file_path,
        },
        in_reply_to=msg_id,
    )


register("pdv.help", handle_help)
register("pdv.tree.resolve_path", handle_resolve_path)

File: pdv-julia/src/handlers/introspection.jl
# handlers/introspection.jl — pdv.help and pdv.tree.resolve_path handlers.
#
# Port of pdv/handlers/introspection.py. Gives the PDV MCP server read-only
# insight into the kernel namespace and file-backed tree nodes.

# Resolve a dotted symbol string to a live Julia object. The head segment is
# looked up in Main first, then among loaded packages (never triggering a
# package load — the analog of Python's sys.modules-only rule).
function _resolve_symbol(symbol::AbstractString)
    isempty(symbol) && throw(ArgumentError("symbol must be a non-empty string"))
    parts = split(symbol, ".")
    head = Symbol(parts[1])
    local obj
    if isdefined(Main, head)
        obj = getfield(Main, head)
    else
        loaded = loaded_module(head)
        if loaded !== nothing
            obj = loaded
        elseif isdefined(@__MODULE__, head)
            # Bare-name fallback: `pdv.help("PDVScript")` should resolve the
            # way `PDVKernel.PDVScript` would.
            obj = getfield(@__MODULE__, head)
        else
            throw(ArgumentError(
                "Could not resolve '$symbol': '$head' is not loaded in the kernel. " *
                "Try the fully qualified name (e.g. 'PDVKernel.$head') or use a " *
                "variable from the namespace. pdv.help does not load packages."))
        end
    end
    for attr in parts[2:end]
        sym = Symbol(attr)
        if obj isa Module || obj isa Type
            isdefined(obj, sym) || throw(ArgumentError(
                "Could not resolve '$symbol': no field '$attr'"))
            obj = getfield(obj, sym)
        else
            obj = getproperty(obj, sym)
        end
    end
    return obj
end

function _object_kind(obj)::String
    obj isa Module && return "module"
    obj isa Type && return "class"
    obj isa Function && return "function"
    return "object"
end

# Signature summary for a function: its method list, capped for readability.
function _signature_of(obj)
    obj isa Function || obj isa Type || return nothing
    ms = methods(obj)
    isempty(ms) && return nothing
    shown = [sprint(show, m) for m in Iterators.take(ms, 5)]
    suffix = length(ms) > 5 ? "\n… ($(length(ms)) methods total)" : ""
    return join(shown, "\n") * suffix
end

function _doc_of(obj)
    try
        doc = Base.Docs.doc(obj)
        text = string(doc)
        # The default "No documentation found" stub is noise, not docs.
        occursin("No documentation found", text) && return nothing
        return text
    catch
        return nothing
    end
end

function _source_of(obj)
    obj isa Function || return nothing
    try
        ms = collect(methods(obj))
        isempty(ms) && return nothing
        file, line = functionloc(ms[1])
        (file === nothing || !isfile(string(file))) && return nothing
        lines = readlines(string(file))
        line > length(lines) && return nothing
        # Best-effort: from the definition line to the matching `end` at the
        # same indentation, capped at 200 lines.
        first_line = lines[line]
        indent = length(first_line) - length(lstrip(first_line))
        stop = min(length(lines), line + 200)
        for i in (line + 1):stop
            stripped = strip(lines[i])
            line_indent = length(lines[i]) - length(lstrip(lines[i]))
            if stripped == "end" && line_indent <= indent
                stop = i
                break
            end
        end
        return join(lines[line:stop], "\n")
    catch
        return nothing
    end
end

"""
    handle_help(msg)

Handle `pdv.help`: resolve a dotted symbol and return kind, signature,
docstring, and (optionally) source.
"""
function handle_help(msg::AbstractDict)
    msg_id = get(msg, "msg_id", nothing)
    payload = get(msg, "payload", Dict{String,Any}())
    payload isa AbstractDict || (payload = Dict{String,Any}())
    symbol = string(get(payload, "symbol", ""))
    include_source = get(payload, "include_source", false) == true

    if isempty(symbol)
        send_error("pdv.help.response", "introspection.missing_symbol",
                   "symbol is required in pdv.help payload"; in_reply_to=msg_id)
        return nothing
    end

    obj = try
        _resolve_symbol(symbol)
    catch err
        send_error("pdv.help.response", "introspection.symbol_not_found",
                   "Could not resolve symbol '$symbol': $(sprint(showerror, err))";
                   in_reply_to=msg_id)
        return nothing
    end

    send_message("pdv.help.response", Dict{String,Any}(
        "symbol" => symbol,
        "kind" => _object_kind(obj),
        "signature" => _signature_of(obj),
        "doc" => _doc_of(obj),
        "source" => include_source ? _source_of(obj) : nothing);
        in_reply_to=msg_id)
    nothing
end

# Recursively collect (tree_path, AbstractPDVFile) pairs from a container.
function _collect_file_nodes(value, tree_path::String,
                             out::Vector{Tuple{String,AbstractPDVFile}})
    if value isa AbstractPDVFile
        push!(out, (tree_path, value))
        return nothing
    end
    if value isa AbstractPDVTree
        for key in collect(keys(value.data))
            child_path = isempty(tree_path) ? key : "$tree_path.$key"
            _collect_file_nodes(value.data[key], child_path, out)
        end
    elseif value isa AbstractDict
        for key in collect(keys(value))
            child_path = isempty(tree_path) ? string(key) : "$tree_path.$(key)"
            _collect_file_nodes(value[key], child_path, out)
        end
    end
    nothing
end

# Extract the <uuid> segment from a path under a tree/ directory.
function _extract_uuid_segment(abs_path::String)
    parts = splitpath(normpath(abs_path))
    for (index, part) in enumerate(parts)
        if part == "tree" && index + 1 <= length(parts)
            candidate = parts[index+1]
            isempty(candidate) || return candidate
        end
    end
    return nothing
end

"""
    handle_resolve_path(msg)

Handle `pdv.tree.resolve_path`: bidirectional translation between a
dot-delimited tree path and an absolute filesystem path.
"""
function handle_resolve_path(msg::AbstractDict)
    msg_id = get(msg, "msg_id", nothing)
    validated = validate_register_request(msg, "pdv.tree.resolve_path.response",
                                          "introspection"; required_fields=("path",))
    validated === nothing && return nothing
    tree, payload = validated
    path = string(get(payload, "path", ""))
    working_dir = tree.working_dir

    if isabspath(path)
        # Reverse: filesystem path -> tree path(s).
        target_real = ispath(path) ? realpath(path) : normpath(path)
        target_uuid = _extract_uuid_segment(target_real)

        file_nodes = Tuple{String,AbstractPDVFile}[]
        _collect_file_nodes(tree, "", file_nodes)

        tree_paths = String[]
        for (node_path, node) in file_nodes
            matched = target_uuid !== nothing && node.uuid == target_uuid
            if !matched && working_dir !== nothing
                node_real = try
                    p = resolve_path(node, working_dir)
                    ispath(p) ? realpath(p) : normpath(p)
                catch
                    nothing
                end
                matched = node_real !== nothing && node_real == target_real
            end
            matched && push!(tree_paths, node_path)
        end

        send_message("pdv.tree.resolve_path.response", Dict{String,Any}(
            "input" => path, "tree_paths" => tree_paths, "file_path" => target_real);
            in_reply_to=msg_id)
        return nothing
    end

    # Forward: tree path -> filesystem path.
    if !haskey(tree, path)
        send_error("pdv.tree.resolve_path.response", "tree.path_not_found",
                   "No node at tree path: '$path'"; in_reply_to=msg_id)
        return nothing
    end
    node = tree[path]
    file_path = nothing
    if node isa AbstractPDVFile
        file_path = try
            resolve_path(node, working_dir)
        catch
            nothing
        end
    end

    send_message("pdv.tree.resolve_path.response", Dict{String,Any}(
        "input" => path, "tree_paths" => [path], "file_path" => file_path);
        in_reply_to=msg_id)
    nothing
end

register_message_handler("pdv.help", handle_help)
register_message_handler("pdv.tree.resolve_path", handle_resolve_path)
